x-pack/plugins/cases/public/common/hooks.ts

```diff
@@ -5,11 +5,97 @@
  * 2.0.
  */
 
+import { useCallback, useEffect, useState } from 'react';
+
+import { useLocation, useHistory } from 'react-router-dom';
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { stringify } from 'query-string';
 import { STACK_APP_ID } from '../../common/constants';
 import { useCasesContext } from '../components/cases_context/use_cases_context';
 
+import { DEFAULT_QUERY_PARAMS } from '../containers/use_get_cases';
+import { parseUrlQueryParams } from './utils';
+
+import type { LocalStorageQueryParams, QueryParams, UrlQueryParams } from '../../common/ui/types';
+
 export const useIsMainApplication = () => {
   const { appId } = useCasesContext();
 
   return appId === STACK_APP_ID;
 };
+
+export function useUrlState(isSelectorView: boolean = false) {
+  const location = useLocation();
+  const history = useHistory();
+
+  const [queryParams, setQueryParams] = useState<UrlQueryParams>({});
```

Comment:
Would it be problematic if we did `useState<...>({sortField: DEFAULT_QUERY_PARAMS.sortField, sortOrder: DEFAULT_QUERY_PARAMS.sortOrder});

I think if we did that we could change the `UrlQueryParams` to alway shave those props defined right? Then they wouldn't need to be defaulted in the `all_cases` component.

Reply:
That would fix it but then I was thinking "why not just `{ ...DEFAULT_QUERY_PARAMS }`?" so I tested it further to remember my initial reasoning.

[The problem is this line.](https://github.com/elastic/kibana/pull/144228/files#diff-e76d7ec20380a26ce9055eb18c05773148decf96628b247f273db0891a19fca8R56)

I added that check because if there are multiple successive renders we don't want to call `history.replace` unnecessarily often(which throws that unsafe operation error we saw before).

Now, if I set `queryParams` to `DEFAULT_QUERY_PARAMS` by default, during the first render when `persistAndUpdateQueryParams` is called the `if` will fail and the URL and local storage won't be updated.